File: uniffi_bindgen/src/bindings/kotlin/templates/macros.kt
{#
// Template to call into rust. Used in several places.
// Variable names in `arg_list` should match up with arg lists
// passed to rust via `arg_list_lowered`
#}

{%- macro to_ffi_call(func) -%}
    {%- match func.self_type() %}
    {%- when Some(Type::Object { .. }) %}
    preventHandleFree {
        {%- call to_raw_ffi_call(func) %}
    }
    {% else %}
        {%- call to_raw_ffi_call(func) %}
    {% endmatch %}
{%- endmacro %}

{%- macro to_raw_ffi_call(func) -%}
    uniffiRun {
        {%- let args = func.full_arguments() %}
        {%- if !args.is_empty() %}
        val uniffiFfiBuffer = Memory(
            max(
                // Size needed for the arguments
                {%- for arg in args.iter() %}
                {{ arg|ffi_serializer_name(ci) }}.size(){% if !loop.last %} + {% endif %}
                {%- endfor -%}
                ,
                // Size needed for the return value
                {% if let Some(return_ty) = func.return_type() %}{{ return_ty|ffi_serializer_name(ci) }}.size() + {% endif -%}
                UniffiFfiSerializerUniffiRustCallStatus.size()
            )
        )
        {%- else %}
        val uniffiFfiBuffer = Memory(
            // Size needed for the return value
            {% if let Some(return_ty) = func.return_type() %}{{ return_ty|ffi_serializer_name(ci) }}.size() + {% endif -%}
            UniffiFfiSerializerUniffiRustCallStatus.size()
        )
        {%- endif %}
        {%- if !args.is_empty() %}
        val uniffiArgsCursor = UniffiBufferCursor(uniffiFfiBuffer)
        {%- for arg in args.iter() %}
        {{ arg|ffi_serializer_name(ci) }}.write(uniffiArgsCursor, {{ arg|lower_fn }}({{ arg|arg_name }}));
        {%- endfor %}
        {%- endif %}

        UniffiLib.{{ func.ffi_func().pointer_ffi_name() }}(uniffiFfiBuffer)
        val uniffiReturnCursor = UniffiBufferCursor(uniffiFfiBuffer)
        val uniffiCallStatus = UniffiFfiSerializerUniffiRustCallStatus.read(uniffiReturnCursor)

        {%- if let Some(throws_ty) = func.throws_type() %}
        {%- if ci.is_external(throws_ty) %}
        uniffiCheckCallStatus({{ throws_ty|type_name(ci) }}ExternalErrorHandler, uniffiCallStatus)
        {%- else %}
        uniffiCheckCallStatus({{ throws_ty|type_name(ci) }}, uniffiCallStatus)
        {%- endif %}
        {%- else %}
        uniffiCheckCallStatus(UniffiNullRustCallStatusErrorHandler, uniffiCallStatus)
        {%- endif %}

        {%- if let Some(return_ty) = func.return_type() %}
        {{ return_ty|ffi_serializer_name(ci) }}.read(uniffiReturnCursor)
        {%- endif %}
    }
{%- endmacro -%}

{%- macro func_decl(func_decl, callable, indent) %}
    {%- call docstring(callable, indent) %}

    {%- match callable.throws_type() -%}
    {%-     when Some(throwable) %}
    @Throws({{ throwable|type_name(ci) }}::class)
    {%-     else -%}
    {%- endmatch -%}
    {%- if callable.is_async() %}
    @Suppress("ASSIGNED_BUT_NEVER_ACCESSED_VARIABLE")
    {{ func_decl }} suspend fun {{ callable.name()|fn_name }}(
        {%- call arg_list(callable, callable.self_type().is_none()) -%}
    ){% match callable.return_type() %}{% when Some(return_type) %} : {{ return_type|type_name(ci) }}{% when None %}{%- endmatch %} {
        {% call call_async(callable) %}
    }
    {%- else -%}
    {{ func_decl }} fun {{ callable.name()|fn_name }}(
        {%- call arg_list(callable, callable.self_type().is_none()) -%}
    ){%- match callable.return_type() -%}
    {%-         when Some(return_type) -%}
        : {{ return_type|type_name(ci) }} {
            return {{ return_type|lift_fn }}({% call to_ffi_call(callable) %})
    }
    {%-         when None %}
        = {% call to_ffi_call(callable) %}
    {%-     endmatch %}
    {% endif %}
{% endmacro %}

{%- macro call_async(callable) -%}
    {%- match callable.self_type() %}
    {%- when Some(Type::Object { .. }) %}
    val uniffiRustFuture = preventHandleFree {
        {%- call to_raw_ffi_call_async(callable) %}
    }
    {% else %}
    val uniffiRustFuture = {%- call to_raw_ffi_call_async(callable) %}
    {% endmatch %}

    val uniffiReturnBufSize = UniffiFfiSerializerUniffiRustCallStatus.size()
        {%- if let Some(return_ty) = callable.return_type() %} + {{ return_ty|ffi_serializer_name(ci) }}.size(){% endif %}

    return uniffiDriveRustFutureToCompletion(uniffiRustFuture, uniffiReturnBufSize) { uniffiFfiBuffer ->
        val uniffiCallStatus = UniffiFfiSerializerUniffiRustCallStatus.read(uniffiFfiBuffer)

        {%- if let Some(throws_ty) = callable.throws_type() %}
        {%- if ci.is_external(throws_ty) %}
        uniffiCheckCallStatus({{ throws_ty|type_name(ci) }}ExternalErrorHandler, uniffiCallStatus)
        {%- else %}
        uniffiCheckCallStatus({{ throws_ty|type_name(ci) }}, uniffiCallStatus)
        {%- endif %}
        {%- else %}
        uniffiCheckCallStatus(UniffiNullRustCallStatusErrorHandler, uniffiCallStatus)
        {%- endif %}

        {%- if let Some(return_ty) = callable.return_type() %}
        {{ return_ty|lift_fn }}(
            {{ return_ty|ffi_serializer_name(ci) }}.read(uniffiFfiBuffer)
        )
        {%- endif %}
    }
{%- endmacro %}

{%- macro to_raw_ffi_call_async(func) -%}
    uniffiRun {
        {%- let args = func.full_arguments() %}
        {%- if !args.is_empty() %}
        val uniffiFfiBuffer = Memory(
            max(
                // Size needed for the arguments
                {%- for arg in args.iter() %}
                {{ arg|ffi_serializer_name(ci) }}.size(){% if !loop.last %} + {% endif %}
                {%- endfor -%}
                ,
                // Size needed for the return value
                {% if let Some(return_ty) = func.return_type() %}{{ return_ty|ffi_serializer_name(ci) }}.size() + {% endif -%}
                UniffiFfiSerializerUniffiRustCallStatus.size()
            )
        )
        {%- else %}
        val uniffiFfiBuffer = Memory(
            // Size needed for the return value
            {% if let Some(return_ty) = func.return_type() %}{{ return_ty|ffi_serializer_name(ci) }}.size() + {% endif -%}
            UniffiFfiSerializerUniffiRustCallStatus.size()
        )
        {%- endif %}

        {%- if !args.is_empty() %}
        val uniffiArgsCursor = UniffiBufferCursor(uniffiFfiBuffer)
        {%- for arg in args.iter() %}
        {{ arg|ffi_serializer_name(ci) }}.write(uniffiArgsCursor, {{ arg|lower_fn }}({{ arg|arg_name }}));
        {%- endfor %}
        {%- endif %}

        UniffiLib.{{ func.ffi_func().pointer_ffi_name() }}(uniffiFfiBuffer)

        // Async functions always return a future handle, regardless of the return value of the
        // actual function
        val uniffiReturnCursor = UniffiBufferCursor(uniffiFfiBuffer)
        UniffiFfiSerializerHandle.read(uniffiReturnCursor)
    }
{%- endmacro -%}


{%- macro arg_list_lowered(func) %}
    {%- for arg in func.arguments() %}
        {{- arg|lower_fn }}({{ arg.name()|var_name }}),
    {%- endfor %}
{%- endmacro -%}

{#-
// Arglist as used in kotlin declarations of methods, functions and constructors.
// If is_decl, then default values be specified.
// Note the var_name and type_name filters.
-#}

{% macro arg_list(func, is_decl) %}
{%- for arg in func.arguments() -%}
        {{ arg.name()|var_name }}: {{ arg|type_name(ci) }}
{%-     if is_decl %}
{%-         match arg.default_value() %}
{%-             when Some(default) %} = {{ default|render_default(arg, ci) }}
{%-             else %}
{%-         endmatch %}
{%-     endif %}
{%-     if !loop.last %}, {% endif -%}
{%- endfor %}
{%- endmacro %}

{#-
// Arglist as used in the UniffiLib function declarations.
// Note unfiltered name but ffi_type_name filters.
-#}
{%- macro arg_list_ffi_decl(func) %}
    {%- for arg in func.arguments() %}
        {{- arg.name()|var_name }}: {{ arg.type_().borrow()|ffi_type_name(ci) -}},
    {%- endfor %}
    {%- if func.has_rust_call_status_arg() %}uniffi_out_err: UniffiRustCallStatus, {% endif %}
{%- endmacro -%}

{% macro field_name(field, field_num) %}
{%- if field.name().is_empty() -%}
v{{- field_num -}}
{%- else -%}
{{ field.name()|var_name }}
{%- endif -%}
{%- endmacro %}

{% macro field_name_unquoted(field, field_num) %}
{%- if field.name().is_empty() -%}
v{{- field_num -}}
{%- else -%}
{{ field.name()|var_name|unquote }}
{%- endif -%}
{%- endmacro %}

 // Macro for destroying fields
{%- macro destroy_fields(member) %}
    Disposable.destroy(
    {%- for field in member.fields() %}
        this.{%- call field_name(field, loop.index) -%}{% if loop.last %}{% else %},{% endif -%}
    {%- endfor %}
    )
{%- endmacro -%}

{%- macro docstring_value(maybe_docstring, indent_spaces) %}
{%- match maybe_docstring %}
{%- when Some(docstring) %}
{{ docstring|docstring(indent_spaces) }}
{%- else %}
{%- endmatch %}
{%- endmacro %}

{%- macro docstring(defn, indent_spaces) %}
{%- call docstring_value(defn.docstring(), indent_spaces) %}
{%- endmacro %}

// macro for uniffi_trait implementations.
{% macro uniffi_trait_impls(uniffi_trait_methods) %}
{# We have 2 display traits, kotlin has 1. Prefer `Display` but use `Debug` otherwise #}
{%- if let Some(fmt) = uniffi_trait_methods.display_fmt.or(uniffi_trait_methods.debug_fmt.clone()) %}
    // The local Rust `Display`/`Debug` implementation.
    override fun toString(): String {
        return {{ fmt.return_type().unwrap()|lift_fn }}({% call to_ffi_call(fmt) %})
    }
{%- endif %}
{%- if let Some(eq) = uniffi_trait_methods.eq_eq %}
    // The local Rust `Eq` implementation - only `eq` is used.
    override fun equals(other: Any?): Boolean {
        if (other !is {{ eq.object_name()|class_name(ci) }}) return false
        return {{ eq.return_type().unwrap()|lift_fn }}({% call to_ffi_call(eq) %})
    }
{%- endif %}
{%- if let Some(hash) = uniffi_trait_methods.hash_hash %}
    // The local Rust `Hash` implementation
    override fun hashCode(): Int {
        return {{ hash.return_type().unwrap()|lift_fn }}({%- call to_ffi_call(hash) %}).toInt()
    }
{%- endif %}
{%- if let Some(cmp) = uniffi_trait_methods.ord_cmp %}
    // The local Rust `Ord` implementation
    override fun compareTo(other: {{ cmp.object_name()|class_name(ci) }}): Int {
        return {{ cmp.return_type().unwrap()|lift_fn }}({%- call to_ffi_call(cmp) %}).toInt()
    }
{%- endif %}
{%- endmacro %}

File: uniffi_bindgen/src/bindings/swift/templates/macros.swift
{#
// Template to call into rust. Used in several places.
// Variable names in `arg_list_decl` should match up with arg lists
// passed to rust via `arg_list_lowered`
#}

{%- macro to_ffi_call(func) -%}
    {%- call is_try(func) -%}
    {%- if let(Some(e)) = func.throws_type() -%}
        rustCallWithError({{ e|ffi_error_converter_name }}_lift) {
    {%- else -%}
        rustCall() {
    {%- endif %}
    {{ func.ffi_func().name() }}(
        {%- match func.self_type() %}
        {%-     when Some(Type::Object { .. }) %}
            self.uniffiCloneHandle(),
        {%-     when Some(t) %}
            {{ t|lower_fn }}(self),
        {%-     when None %}
        {%- endmatch %}
        {%- call arg_list_lowered(func) -%} $0
    )
}
{%- endmacro -%}

// eg, `public func foo_bar() { body }`
{%- macro func_decl(func_decl, callable, indent) %}
{%- call docstring(callable, indent) %}
{{ func_decl }} {{ callable.name()|fn_name }}(
    {%- call arg_list_decl(callable) -%})
    {%- call is_async(callable) %}
    {%- call throws(callable) %}
    {%- if let Some(return_type) = callable.return_type() %} -> {{ return_type|type_name }} {%- endif %}  {
    {%- call call_body(callable) %}
}
{%- endmacro %}

// primary ctor - no name, no return-type.
{%- macro ctor_decl(callable, indent) %}
{%- call docstring(callable, indent) %}
public convenience init(
    {%- call arg_list_decl(callable) -%}) {%- call is_async(callable) %} {%- call throws(callable) %} {
    {%- if callable.is_async() %}
    let handle =
        {%- call call_async(callable) %}
        {# The async mechanism returns an already constructed self.
           We work around that by cloning the handle from that object, then
           assume the old object dies as there are no other references possible.
        #}
        .uniffiCloneHandle()
    {%- else %}
    let handle =
        {% call to_ffi_call(callable) %}
    {%- endif %}
    self.init(unsafeFromHandle: handle)
}
{%- endmacro %}

{%- macro call_body(callable) %}
{%- if callable.is_async() %}
    return {%- call call_async(callable) %}
{%- else %}
{%-     match callable.return_type() -%}
{%-         when Some(return_type) %}
    return {% call is_try(callable) %} {{ return_type|lift_fn }}({% call to_ffi_call(callable) %})
{%-         when None %}
{%-             call to_ffi_call(callable) %}
{%-     endmatch %}
{%- endif %}

{%- endmacro %}

{%- macro call_async(callable) %}
        {% call is_try(callable) %} await uniffiRustCallAsync(
            rustFutureFunc: {
                {{ callable.ffi_func().name() }}(
                    {%- if callable.self_type().is_some() %}
                    self.uniffiCloneHandle(){% if !callable.arguments().is_empty() %},{% endif %}
                    {% endif %}
                    {%- for arg in callable.arguments() -%}
                    {{ arg|lower_fn }}({{ arg.name()|var_name }}){% if !loop.last %},{% endif %}
                    {%- endfor %}
                )
            },
            pollFunc: {{ callable.ffi_rust_future_poll(ci) }},
            completeFunc: {{ callable.ffi_rust_future_complete(ci) }},
            freeFunc: {{ callable.ffi_rust_future_free(ci) }},
            {%- match callable.return_type() %}
            {%- when Some(return_type) %}
            liftFunc: {{ return_type|lift_fn }},
            {%- when None %}
            liftFunc: { $0 },
            {%- endmatch %}
            {%- match callable.throws_type() %}
            {%- when Some with (e) %}
            errorHandler: {{ e|ffi_error_converter_name }}_lift
            {%- else %}
            errorHandler: nil
            {% endmatch %}
        )
{%- endmacro %}

{%- macro arg_list_lowered(func) %}
    {%- for arg in func.arguments() %}
        {{ arg|lower_fn }}({{ arg.name()|var_name }}),
    {%- endfor %}
{%- endmacro -%}

{#-
// Arglist as used in Swift declarations of methods, functions and constructors.
// Note the var_name and type_name filters.
-#}

{% macro arg_list_decl(func) %}
    {%- for arg in func.arguments() -%}
        {% if config.omit_argument_labels() %}_ {% endif %}{{ arg.name()|var_name }}: {{ arg|type_name -}}
        {%- match arg.default_value() %}
        {%- when Some with(default) %} = {{ default|default_swift(arg) }}
        {%- else %}
        {%- endmatch %}
        {%- if !loop.last %}, {% endif -%}
    {%- endfor %}
{%- endmacro %}

{#-
// Field lists as used in Swift declarations of Records and Enums.
// Note the var_name and type_name filters.
-#}
{% macro field_list_decl(item, has_nameless_fields) %}
    {%- for field in item.fields() -%}
        {%- call docstring(field, 8) %}
        {%- if has_nameless_fields %}
        {{- field|type_name -}}
        {%- if !loop.last -%}, {%- endif -%}
        {%- else -%}
        {{ field.name()|var_name }}: {{ field|type_name -}}
        {%- match field.default_value() %}
            {%- when Some with(default) %} = {{ default|default_swift(field) }}
            {%- else %}
        {%- endmatch -%}
        {% if !loop.last %}, {% endif %}
        {%- endif -%}
    {%- endfor %}
{%- endmacro %}

{% macro field_name(field, field_num) %}
{%- if field.name().is_empty() -%}
v{{- field_num -}}
{%- else -%}
{{ field.name()|var_name }}
{%- endif -%}
{%- endmacro %}

{% macro arg_list_protocol(func) %}
    {%- for arg in func.arguments() -%}
        {% if config.omit_argument_labels() %}_ {% endif %}{{ arg.name()|var_name }}: {{ arg|type_name -}}
        {%- if !loop.last %}, {% endif -%}
    {%- endfor %}
{%- endmacro %}

{%- macro is_async(func) %}
{%- if func.is_async() %}async {% endif %}
{%- endmacro -%}

{%- macro throws(func) %}
{%- if func.throws() %}throws {% endif %}
{%- endmacro -%}

{%- macro is_try(func) %}
{%- if func.throws() %}try {% else %}try! {% endif %}
{%- endmacro -%}

{%- macro docstring_value(maybe_docstring, indent_spaces) %}
{%- match maybe_docstring %}
{%- when Some(docstring) %}
{{ docstring|docstring(indent_spaces) }}
{%- else %}
{%- endmatch %}
{%- endmacro %}

{%- macro docstring(defn, indent_spaces) %}
{%- call docstring_value(defn.docstring(), indent_spaces) %}
{%- endmacro %}

// macro for uniffi_trait implementations.
{% macro uniffi_trait_impls(uniffi_trait_methods) %}
{%- if let Some(fmt) = uniffi_trait_methods.debug_fmt %}
// The local Rust `Debug` implementation.
extension {{ fmt.object_name() }}: CustomDebugStringConvertible {
    public var debugDescription: String {
        return {% call is_try(fmt) %} {{ fmt.return_type().unwrap()|lift_fn }}(
            {% call to_ffi_call(fmt) %}
        )
    }
}
{%- endif %}
{%- if let Some(fmt) = uniffi_trait_methods.display_fmt %}
// The local Rust `Display` implementation.
extension {{ fmt.object_name() }}: CustomStringConvertible {
    public var description: String {
        return {% call is_try(fmt) %} {{ fmt.return_type().unwrap()|lift_fn }}(
            {% call to_ffi_call(fmt) %}
        )
    }
}
{%- endif %}
{%- if let Some(eq) = uniffi_trait_methods.eq_eq %}
// The local Rust `Eq` implementation - only `eq` is used.
extension {{ eq.object_name() }}: Equatable {
    public static func == (self: {{ eq.object_name() }}, other: {{ eq.object_name() }}) -> Bool {
        return {% call is_try(eq) %} {{ eq.return_type().unwrap()|lift_fn }}(
            {% call to_ffi_call(eq) %}
        )
    }
}
{%- endif %}
{%- if let Some(hash) = uniffi_trait_methods.hash_hash %}
// The local Rust `Hash` implementation
extension {{ hash.object_name() }}: Hashable {
    public func hash(into hasher: inout Hasher) {
        let val = {% call is_try(hash) %} {{ hash.return_type().unwrap()|lift_fn }}(
            {% call to_ffi_call(hash) %}
        )
        hasher.combine(val)
    }
}
{%- endif %}
{%- if let Some(cmp) = uniffi_trait_methods.ord_cmp %}
// The local Rust `Ord` implementation
extension {{ cmp.object_name() }}: Comparable {
    public static func < (self: {{ cmp.object_name() }}, other: {{ cmp.object_name() }}) -> Bool {
        return {% call is_try(cmp) %} {{ cmp.return_type().unwrap()|lift_fn }}(
            {% call to_ffi_call(cmp) %}
        ) < 0
    }
}
{%- endif %}
{%- endmacro %}
